### ISSUE TITLE
Configurations questions

### ISSUE THREAD

**Issue by JoomFX** [gantry/gantry5#213]
@w00fz @mahagr @arifinrt  - first of all sorry for the long "essay" and feel free to close it directly if you find it ridiculous. I was thinking a lot if I should even post this...

so, as I said earlier, in the Gitter chat, I'd like to discuss something with you guys. I just cannot stop thinking about it since Beta 1 was released.
I'd like to share my point of view and how I'm planning to develop my G5-based templates. I would really, really appreciate if you can share your thoughts and if we can make a nice discussion about it as I really want to understand the "Configurations" concept (I do understand it but you will see what I mean when you read what I will write below)...

So, let's start... I installed the Callisto Rocketlauncher and I started playing with it. The thing that got my attention first (I was expecting it and I wrote about it when Beta 1 was released) was the number of Configurations/Styles you have in the Template Manager. There are 16 Configurations (for more complicated designs or templates with more features/pages we can easily reach 30, 40...)

I do understand that you want to keep everything in G5 so it is CMS agnostic. By doing so you'll be able to convert one design for different CMSs in minutes/hours. You probably develop the Templates/Themes on Prime and then you port them to the particular CMS (currently only Joomla). So far so good...

In my opinion, having so many Configurations might be a bit confusing for the user. Me and you can work perfectly fine with it, (I'm playing with G5 for more than a month now and you are the developers :D ) but imagine how the old school Joomla user will react when he sees so many Configurations. It is not only the old school user, other users as well, and it is not only the number of Configurations but also the way the Particles (these awesome Particles, thanks @mahagr ) are being assigned to a Configuration, not to a Menu Item as everything else in Joomla.

So, here is how I'm planning to develop/configure my Joomla Templates... using one, just one Configuration (Template_Name - Default):

1. I will create a default Layout (YAML) that I will use as a basis for most of my designs. This layout will have about 10 sections (including left and right sidebar that will collapse if empty) and all of them will be included under the Styles tab so the user will be able to apply different styles to each section. Of course, not all of these 10 sections will be used in the template, but the user will be able to move a module position to a different section (some of the sections will be empty) and style it as he likes. (btw, in the Callisto Template, some sections are missing from the Styles Tab - `mainfeature` ,  `sidefeature` , `mainbar` and `sidebar` but probably it is intentional ).

2. If I do not want to have "Page Content" (Mainbody in G4) loaded on the Homepage, I will use the "G5 Themes -> Custom Page" menu type. This eliminates the need of having a separate Configuration for the Homepage.

3. I will configure and publish all my Particles from the Module Manager using this - https://github.com/gantry/gantry5/issues/163 . Hopefully it will be implemented soon. No rush at all, I do understand you have different priorities.

So basically, that's it. Using just one Configuration. Of course, as it was in G4, the user might create as many as he wants, or the Template Developer can create as many as he wants for more complicated designs, etc.

Using the above method, especially point 3, makes the Template a bit CMS dependent, because of using the Module Manager. But this should not be a big deal, because I suppose you will have the same Wordpress Widget, so you can just copy/paste the setting from a Joomla Particle Module into it. And for Grav... not sure :)

So, in short:

1. What do you think of my idea, my way of developing Templates? There will be just one Configuration and just one layout under the "Load" button.
2. Why do you think it is better to have all these Configurations?

**Comment by mahagr** [comment]
Well, we did notice the number of configurations as well, but right now we have no means to fight against configuration creep, mostly because of we do not have particles module, which would fix most of the issues. Every of those pages have 1 main content area and then 2 particles that were changing.

It was something we really didn't consider in time, though we had our hands full on other parts of the code as well. When we discovered it, we just decided that we have to live with the excess number of configurations for this template and try to fix the main issue as soon as we have some room to breathe.

Basically the goal is to have just few layouts to work with and do the rest by using either component content or module positions. Different configurations are mainly meant for different parts of the site, so you can set unique style or different look to make those stand out.

Basically the most important thing to archive this is to use module positions in joomla and to implement #163 but I'd like to go a bit further at some point and allow adding particles for the main content as well.

**Comment by JoomFX** [comment]
@mahagr  thank you very much for the detailed answer mate :)
I'm really happy because of what I just read above:

*Different configurations are mainly meant for different parts of the site, so you can set unique style or different look to make those stand out*

Yes, exactly! As it was in G4. 
I'm really happy that we think alike. Website configuration should be done mostly in the CMS, while the style/design should be done in Gantry5 :)

Of course guys, of course you cannot do everything overnight. And nobody expects that - we all see how hard you work! And I wish I had a better grasp of PHP so I could just dive into it and write some of the code for you :)

And guys, would you please give me your opinion on the following, as I'm still a bit confused with it - the Layouts/Presets under the "Load" button.
While the "Platform" ones might ease the transition from a Protostar or Gantry4 based website to a Gantry5 Template, why would you need different "User" layouts/presets? I mean why would you have a preset with Sidebar on the left, another one with Sidebar on the right, another one with different arrangement of the Sections and so on, instead of just having one "Universal" Preset that gives you both Sidebar on the Left, Sidebar on right and just enough sections to do whatever you want?
Probably I'm missing something but currently I do not understand what is the point to have multiple Layouts/Presets :)

**Comment by w00fz** [comment]
I agree with you, I think a maximum of 3-4 Configurations can probably cover an entire site. I think having so many Configurations is not necessary a bad thing but it is something that should grow over time.
If i start building a theme for myself from scratch, I might create 2 configurations and then overtime I might add new ones to cover some special requirements  I want out of a page. Although one thing is starting with a few Configurations and having them grow over time, another one is starting with so many right away: it can be overwhelming.

I think we all agree with this and we are going to revisit this approach.

Although one thing I would like to highlight is that Callisto configurations are so many only in the RocketLauncher, which like always is meant to be a demo for people to get started with, by no means it is advised nor recommended to be used as-is for a production site, that's how we decided to showcase the template and highlighted all the possible combinations that come with it but in the end it should be customized and trimmed down and nobody should be afraid of deleting Configurations to keep them down. In fact, if you just install the standalone template you will only find Home configuration.

**Comment by JoomFX** [comment]
thanks @w00fz :)

Yes, the Configurations are definitely not a bad thing, everything behind Gantry5 is simply awesome, I'm just trying to figure out the best way to develop a template which will be very flexible and at the same time very intuitive and user friendly, without overcomplicating its configuration/setup.

Really happy we all agree here and the approach will be revisited :)

I don't yet know how the multilingual websites will need to be configured in G5, but if it is as in G4 (which for me is just fine), then imagine a website that by default has 20 Configurations. If you add two additional languages, the Configurations will go to 60 :)

Anyway, no need to dig into it further. I just wanted to discuss it with you because as I said, I couldn't stop thinking about it since Beta 1 was released.

And @w00fz and @mahagr , what do you think about my other question - the Layouts/Presets under the "Load" button. What do you think about having just one "Universal" Layout?

**Comment by mahagr** [comment]
I wouldn't have just a single universal layout, mostly because of the more positions we add, the slower the site will get. So I'd rather have a few different layouts to choose from and the particles being copied over when you change the preset layout.

We are also aware of the issue on multi-language sites and personally I do not agree with Joomla approach, where you basically double everything just to get another language. We do not have that multi-language support implemented yet, but the plan is to be able to translate a single configuration to all your languages.

Again, the reason why Callisto has so many configurations (and presets for all of them) is because of we really weren't fully aware of the implications of not having a particle module. Without modules that you can assign to a page, there's really no other way to implement all the sample content.

**Comment by JoomFX** [comment]
Alright guys, you answered all my questions, again, and I'm very thankful for that :)

If @w00fz does not have anything to add for the "Universal" layout, I think you can go ahead and close this issue ;)

**Comment by sural98** [comment]
Think of a 32-language site. (LTR+RTL)
I am still trying to figure out how I can accomplish such a task with G5 and could not find the best way to the moment.

**Comment by mahagr** [comment]
There is no best way at this moment as we haven't yet had time to look into it. It's something we need for stable version, though, so stay tuned..

**Comment by sural98** [comment]
While beta stage I had assumed that G5 for a multi-language Joomla would be layout base rather than configuration.
Currently we first create many configurations: Home, other pages, languages, then we need to create layouts for every specific configuration.
In a layout based environment we may need just 1 default configuration.
If the base was layout than instead of storing configuration assignments in #_menu layout assignments would be stored in layout file also.
Layouts can then be called using a language filter from within layout page.

**Comment by w00fz** [comment]
Closing this one, hopefully we answered all your questions.
We have other tickets regarding the multi-languages and we know that is still an issue that we haven't yet covered.
